go/batch-submitter/drivers/sequencer/driver.go

```diff
@@ -0,0 +1,203 @@
+		cfg.L1Client,
+	)
+
+	walletAddr := crypto.PubkeyToAddress(cfg.PrivKey.PublicKey)
+
+	return &Driver{
+		cfg:            cfg,
+		ctcContract:    ctcContract,
+		rawCtcContract: rawCtcContract,
+		walletAddr:     walletAddr,
+		ctcABI:         ctcABI,
+	}, nil
+}
+
+// Name is an identifier used to prefix logs for a particular service.
+func (d *Driver) Name() string {
+	return d.cfg.Name
+}
+
+// WalletAddr is the wallet address used to pay for batch transaction fees.
+func (d *Driver) WalletAddr() common.Address {
+	return d.walletAddr
+}
+
+// GetBatchBlockRange returns the start and end L2 block heights that need to be
+// processed. Note that the end value is *exclusive*, therefore if the returned
+// values are identical nothing needs to be processed.
+func (d *Driver) GetBatchBlockRange(
+	ctx context.Context) (*big.Int, *big.Int, error) {
+
+	blockOffset := new(big.Int).SetUint64(d.cfg.BlockOffset)
+
+	start, err := d.ctcContract.GetTotalElements(&bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	start.Add(start, blockOffset)
+
+	latestHeader, err := d.cfg.L2Client.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Add one because end is *exclusive*.
+	end := new(big.Int).Add(latestHeader.Number, bigOne)
+
+	if start.Cmp(end) > 0 {
+		return nil, nil, fmt.Errorf("invalid range, "+
+			"end(%v) < start(%v)", end, start)
+	}
+
+	return start, end, nil
+}
+
+// SubmitBatchTx transforms the L2 blocks between start and end into a batch
+// transaction using the given nonce and gasPrice. The final transaction is
+// published and returned to the call.
+func (d *Driver) SubmitBatchTx(
+	ctx context.Context,
+	start, end, nonce, gasPrice *big.Int) (*types.Transaction, error) {
+
+	name := d.cfg.Name
+
+	log.Info(name+" submitting batch tx", "start", start, "end", end,
+		"gasPrice", gasPrice)
+
+	var blocks []*l2types.Block
+	for i := new(big.Int).Set(start); i.Cmp(end) < 0; i.Add(i, bigOne) {
+		block, err := d.cfg.L2Client.BlockByNumber(ctx, i)
+		if err != nil {
+			return nil, err
+		}
+
+		blocks = append(blocks, block)
+
+		// TODO(conner): remove when moving to multiple blocks
+		break //nolint
+	}
+
+	var batchElements = make([]BatchElement, 0, len(blocks))
+	for _, block := range blocks {
+		batchElements = append(batchElements, BatchElementFromBlock(block))
+	}
+
+	shouldStartAt := start.Uint64()
+	batchParams, err := GenSequencerBatchParams(
+		shouldStartAt, d.cfg.BlockOffset, batchElements,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info(name+" batch params", "params", fmt.Sprintf("%#v", batchParams))
+
+	batchArguments, err := batchParams.Serialize()
+	if err != nil {
+		return nil, err
+	}
+
+	appendSequencerBatchID := d.ctcABI.Methods[appendSequencerBatchMethodName].ID
+	batchCallData := append(appendSequencerBatchID, batchArguments...)
+
+	if uint64(len(batchCallData)) > d.cfg.MaxTxSize {
+		panic("call data too large")
```

Comment:
You are probably aware of this, but when you move to posting more than a single batch element per batch this could end up in an infinite crash loop